File: swift/Sources/WikimediaMetricsPlatform/Curation/Rules/CollectionCurationRules.swift
import Foundation

/**
 * Curation rules to be used with context values with collection types.
 */
struct CollectionCurationRules<T: Hashable & Decodable>: Decodable {
    let contains: T?
    let doesNotContain: T?
    let containsAll: [T]?
    let containsAny: [T]?

    init(
            contains: T? = nil,
            doesNotContain: T? = nil,
            containsAll: [T]? = nil,
            containsAny: [T]? = nil
    ) {
        self.contains = contains
        self.doesNotContain = doesNotContain
        self.containsAll = containsAll
        self.containsAny = containsAny
    }

    func apply(to value: [T]) -> Bool {
        if contains != nil && !value.contains(contains!) {
            return false
        }
        if doesNotContain != nil && value.contains(doesNotContain!) {
            return false
        }
        if containsAll != nil {
            for el in containsAll! {
                if !value.contains(el) {
                    return false
                }
            }
        }
        if containsAny != nil {
            var found = false
            for el in containsAny! {
                if value.contains(el) {
                    found = true
                    break
                }
            }
            if !found {
                return false
            }
        }

        return true
    }

    enum CodingKeys: String, CodingKey {
        case contains
        case doesNotContain = "does_not_contain"
        case containsAll = "contains_all"
        case containsAny = "contains_any"
    }
}

File: java/src/main/java/org/wikimedia/metrics_platform/curation/rules/CollectionCurationRules.java
package org.wikimedia.metrics_platform.curation.rules;

import java.util.Collection;

import com.google.gson.annotations.SerializedName;

import lombok.Builder;
import lombok.EqualsAndHashCode;
import lombok.ToString;

@Builder @EqualsAndHashCode @ToString
public class CollectionCurationRules<T> {
    private T contains;
    @SerializedName("does_not_contain") private T doesNotContain;
    @SerializedName("contains_all") private Collection<T> containsAll;
    @SerializedName("contains_any") private Collection<T> containsAny;

    @SuppressWarnings("checkstyle:CyclomaticComplexity")
    public boolean apply(Collection<T> value) {
        if (contains != null && !value.contains(contains)) {
            return false;
        }
        if (doesNotContain != null && value.contains(doesNotContain)) {
            return false;
        }
        if (containsAll != null && !value.containsAll(containsAll)) {
            return false;
        }
        if (containsAny != null) {
            boolean found = false;
            for (T el : containsAny) {
                if (value.contains(el)) {
                    found = true;
                    break;
                }
            }
            if (!found) {
                return false;
            }
        }
        return true;
    }
}
